File: 0986. Interval List Intersections/main0986.py
from typing import List


class Solution0986:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        res = []
        i, j = 0, 0
        m, n = len(firstList), len(secondList)
        while i < m and j < n:
            left, right = max(firstList[i][0], secondList[j][0]), min(firstList[i][1], secondList[j][1])
            if left <= right:
                res.append([left, right])
            if firstList[i][1] < secondList[j][1]:
                i += 1
            else:
                j += 1
        return res

sol = Solution0986()
firstList = [[0,2],[5,10],[13,23],[24,25]]
secondList = [[1,5],[8,12],[15,24],[25,26]]
res = sol.intervalIntersection(firstList, secondList)
print(res)

File: 0986. Interval List Intersections/main0986.java
import java.util.*;

public class main0986 {
    public static void main(String[] args) {
        Solution0986 sol = new Solution0986();
        int[][] firstList = {{0,2},{5,10},{13,23},{24,25}};
        int[][] secondList = {{1,5},{8,12},{15,24},{25,26}};
        int[][] res = sol.intervalIntersection(firstList, secondList);
        System.out.println(Arrays.deepToString(res));
    }
}

class Solution0986 { // 42 - 47
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        List<int[]> res = new ArrayList<>();
        int i = 0, j = 0;
        int m = firstList.length, n = secondList.length;
        while (i < m && j < n) {
            int left = Math.max(firstList[i][0], secondList[j][0]);     
            int right = Math.min(firstList[i][1], secondList[j][1]);
            if (left <= right) {
                res.add(new int[]{left, right});
            }
            if (firstList[i][1] < secondList[j][1]) {
                i++;
            } else {
                j++;
            }
        }
        return res.toArray(new int[res.size()][]);
    }
}